telegram/ext/_baseupdateprocessor.py

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env python
+#
+# A library that provides a Python interface to the Telegram Bot API
+# Copyright (C) 2015-2023
+# Leandro Toledo de Souza <devs@python-telegram-bot.org>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser Public License for more details.
+#
+# You should have received a copy of the GNU Lesser Public License
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+"""This module contains the BaseProcessor class."""
+from abc import abstractmethod
+from asyncio import BoundedSemaphore
+from types import TracebackType
+from typing import Any, Awaitable, Optional, Type
+
+from telegram.ext import Application
+
+
+class BaseUpdateProcessor:
+    """An abstract base class for update processors. You can use this class to implement
+    your own update processor.
+
+    .. seealso:: :wiki:`Concurrency`
+
+    .. versionadded:: NEXT.VERSION
+
+    Args:
+        max_concurrent_updates (:obj:`int`): The maximum number of updates to be processed
+        concurrently. If this number is exceeded, new updates will be queued until the number
+        of currently processed updates decreases.
```

Comment:
speaking of.. do you want to provide a way to inspect current number of processed updates? (https://github.com/python-telegram-bot/python-telegram-bot/issues/3509#issuecomment-1400636217) can be done by inspecting a property in semaphore itself iirc